testsuite/demo/run-pass/library/pool.verona

```diff
@@ -0,0 +1,235 @@
+    result
+  }
+
+  /*
+   * Return a view on the Cells contents.
+   */
+  use(self: mut): T & iso where return in self {
+    match (self._cell) {
+      // FIXME: The None case should never happen as the cell is only
+      // extracted in the finaliser, an error should really occur
+      // here if it does.
+      // var n: None => {},
+      var c: Cell[T] => (mut-view c).read(),
+    }
+  }
+
+  /*
+   * Finaliser returns the Cell in the Resource to the
+   * original Pool
+   */
+  final(self: mut) {
+    match (self._cell = Builtin.freeze(new None)) {
+      // FIXME: The only time the Cell is extracted is in the finaliser
+      // case, if the contents is None then an error should occur.
+      // var _: None => {}
+      var c: Cell[T] => Pool.add(self._origin, c),
+    }
+  }
+}
+
+/*
+ * A Group is a wrapper around a Queue[Resource[T]], this exists
+ * mostly to provide a nicer interface/integration with the Pools
+ * and Resources.
+ */
+class Group[T] {
+  _resources: Queue[Resource[T]] & iso;
+
+  create(resources: Queue[Resource[T]] & iso): Group[T] & iso {
+    var result = new Group;
+    result._resources = resources;
+    result
+  }
+
+  size(self: mut): U64 & imm {
+    (self._resources).length
+  }
+
+  /*
+   * Get the list of resources in this Group
+   */
+  extract(self: mut): Queue[Resource[T]] & iso {
+    // FIXME: Perhaps this should be swapped with None & imm and
+    // error if one tries to extract the Queue twice
+    self._resources = Queue.create(0)
+  }
+}
+
+/*
+ * A Pool captures a collection of items or Resources of the same type.
+ * Requests can be made to Pool to get n of the Resources.
+ * Requests are serviced in the order that they are received.
+ */
+class Pool[T] {
+  _resources: Queue[Cell[T]] & iso;
+  _requests: Queue[Request[T]] & iso;
+
+  create(): cown[Pool[T]] & imm {
+    var result = new Pool;
+    result._resources = Queue.create(0);
+    result._requests = Queue.create(0);
+    cown.create(result)
+  }
+
+  add(pool: cown[Pool[T]] & imm, resource: Cell[T] & iso) {
+    when (var p = pool) {
+      (mut-view (p._resources)).add(resource);
+      Pool._check_requests(pool);
+    }
+  }
+
+  /*
+   * Give the compiler some help so that it doesn't trip up inferring
+   * the type of the loop when creating the sublist.
+   */
+  _give_resource(pool: cown[Pool[T]] & imm,
+                 resource: (Cell[T] & iso) | (None & imm),
+                 out: Queue[Resource[T]] & mut) {
+    match resource {
+      // The None case should never happen as we already
+      // checked that there are enough resources in the Pool.
+      var _: None => {}
+      var c: Cell[T] => out.add(Resource.create(pool, c)),
+    };
+  }
+
+  /*
+   * Check the first request in the queue, if it can be satisified then
+   * fulfill the requests and schedule a behaviour to check the next
+   * waiting request.
+   */
+  _check_requests(pool: cown[Pool[T]] & imm) {
+    when (var p = pool) {
```

Comment:
So all calls to `_check_requests` are when the pool is already owned by the caller.  Is it important to repeatedly acquire the `pool` here.  Does this improve fairness?  It seems like it is creating more `when`s without improving anything algorithmically.